### PR TITLE
Initial version of new property page documentation

### PR DESCRIPTION
**Pull request by tmeschter**
Add a document describing the architecture of the new project property
pages at a high level, along with a layer diagram.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/project-system/pull/6724)

### CODE REVIEW

**Comment by tmeschter** [comment]
As a follow up to yesterday's conversation in the team meeting about the new property pages, I thought I would draft a document describing, at a high level, all the important pieces and how they relate.

Note that the GitHub diff view doesn't show it, but when you view the .md file the .svg displays in-line.

**Review by jjmew — APPROVED** [review]
:shipit:

File: docs/repo/property-page-architecture.drawio.svg [hunk 1]
```diff
@@ -0,0 +1,234 @@
...like Gecko) Code/1.50.1 Chrome/83.0.4103.122 Electron/9.2.1 Safari/537.36&quot; version=&quot;13.6.5&quot; etag=&quot;olyi-gRlhe6lnlsRP_5v&quot;&gt;&lt;diagram id=&quot;--xkG6ouKXTrO3x_hQlo&quot; name=&quot;Page-1&quot;&gt;5Zhbk9ogGIZ/jZftQDAeLjXutp3pztg6bfeWETS0JF9KiIf++oISjRI7O7OJ2tabhBcw8Lwfxw6Jks07RbP4CRiXnQCxTYdMOkGAQ9w1D6tsnYLDYK8slWBOOwoz8Ys7ETm1EIznJwU1gNQiOxXnkKZ8rk80qhSsT4stQJ5+NaNL7gmzOZW++k0wHe/VLkJH/T0Xy9h9OSwzElqWdUIeUwbrikQeOiRSAHr/lmwiLi28Esu+3uOF3EO7FE/1Syo47Csqi7Jrb036ywfXPL0tu6ygSBm31VCHjNex0HyW0bnNXRuXjRbrRJoUNq8LIWUEEtSuLulRhBYLo+dawQ9eySF9hKLI1oBUV/TF7md01z6uNN9c7CM+kDMhxyHhWm1NkU1py77G9jS5PhpHSjviimllOepiZXn44yNP8+KQ1uMNPbyBxTtV8N2GZYA+FXzX1NH0HyY+uCLxnke8+wfiSCSZ5InpFtUC0nuxwMIXZrYZSbFMTV4iGLNNasEb3L+iOX3PHGLNGVs+ymA27LlaibmZ2htzIoXUkmM0j3fVcTMYgzOMpAZjrwYjaQDjwMMYWowzTVNGlcX40wU5o5rabilYmQVTNYj1rwnww56h4gwOWwrwoedM380+menSttxZXMkghELEaw1CCE/GtzIoQC9wqG7sNOFQ+a2KRT1r0VeqBBT50ZIcCtXGRHSRbNMz1Pk4CK84DjD2KA8qA0FYrijf5pon/8OU5EV8nRetRTzxvMC7Tf7z6OmjeXwuZJNh/rp5R4HbiZHJEDUE/2wgkJqlGre14ylPuhX4w7O1OqssDve2TrdgxyHOb2KHfyDDyPoRFbmG5Bpu3NngQLd0wz+sRdOZERTP4F6GQPMrMRnWMEZtMfbPXAx0ys1R+NEEOLen4jflStw8+FdEe/PguzVXP02BN8njrd0ur3L3SR5+Aw==&lt;/diagram&gt;&lt;/mxfile&gt;">
```
Comment (swesonga) [inline docs/repo/property-page-architecture.drawio.svg:1]: How did you generate the SVG?

**Review by swesonga — APPROVED** [review]

File: docs/repo/property-page-architecture.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+### 9. Standard property providers
+
+CPS has an implementation of `IProjectPropertiesProvider` that handles reading and writing data to the MSBuild project file (the `ProjectPropertiesProvider`), among others. As many of the properties that we change through the project properties UI are in fact stored in the project file, we may heavy (though implicit) use of this type. Generally we only need to think about this type when we run up against its limitations, at which point we need to think about custom `IProjectPropertiesProvider`s.
```
Comment (drewnoakes) [inline docs/repo/property-page-architecture.md:64]:
```suggestion
CPS has an implementation of `IProjectPropertiesProvider` that handles reading and writing data to the MSBuild project file (the `ProjectPropertiesProvider`), among others. As many of the properties that we change through the project properties UI are in fact stored in the project file, we make heavy (though implicit) use of this type. Generally we only need to think about this type when we run up against its limitations, at which point we need to think about custom `IProjectPropertiesProvider`s.
```

File: docs/repo/property-page-architecture.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+### 11. XAML Rules
+
+At the very bottom of the implementation stack are XAML Rule files, so-called because they deserialize to instances of the `Rule` type. This is not to be confused with the `IRule` interface; a `Rule` represents a "page" of properties in the abstract, while `IRule` represents that page bound to a particular project or item.
+
+These Rule files are where the declarative, metadata-driven aspect of the new property pages comes into play. These files declare page, the page's metadata, it's properties, the storage mechanisms to use (i.e. which `IProjectPropertiesProvider` will handle a particular property), and a bunch of other property metadata: documentation links, search terms, descriptions, the property type, whether the value can vary across project configurations, etc.
```
Comment (drewnoakes) [inline docs/repo/property-page-architecture.md:74]:
```suggestion
These Rule files are where the declarative, metadata-driven aspect of the new property pages comes into play. These files declare page, the page's metadata, its properties, the storage mechanisms to use (i.e. which `IProjectPropertiesProvider` will handle a particular property), and a bunch of other property metadata: documentation links, search terms, descriptions, the property type, whether the value can vary across project configurations, etc.
```

**Review by drewnoakes — APPROVED** [review]
Looks great.

We will add client-specific documentation to this in time, detailing editor and other metadata specifications, and walking through various scenarios.

File: docs/repo/property-page-architecture.drawio.svg (hunk 1 above)
Comment (drewnoakes) [reply docs/repo/property-page-architecture.drawio.svg:1]: I'm guessing it can be opened in https://app.diagrams.net/ (part of https://drawio-app.com/)

Comment (tmeschter) [reply docs/repo/property-page-architecture.drawio.svg:1]: That's correct. Specifically, I used the "Draw.io Integration" extension for VS Code. And by saving it as a .drawio.svg file, I can embed it directly into the .md file.

**Comment by tmeschter** [comment]
I'm going to leave this PR open for another day or two to collect feedback. If you have questions about any part of the described design please leave a comment in the PR, and I'll update the docs.

**Review by dmonroym — APPROVED** [review]

**Pull request merged**